Lazily evaluate tracer arguments

lpereira opened the pull request:
Individual commit messages will have more detail, but here's the gist:

> Introduce TRACE_ERROR(), TRACE_WARNING(), TRACE_INFO(), and TRACE_VERBOSE() macros, that supersede similarly-named functions in the trace namespace.
> 
> Using these macros will ensure that the arguments are only evaluated in the expected verbosity levels.  Some arguments create temporary objects that are only visible in high verbosity levels, creating useless churn in the allocator.

This has been suggested by @vitek-karas and is my first attempt at this:

> I did some rough profiling of hostpolicy trying to figure out why some tests take so long to run (1 sec per iteration). And one of the things which jumped out was `version_t::as_str()` which is basically only used for tracing. The short version of the story is that we call something like `trace::verbose("..", version.as_str())`. The compiler can't optimize the call to `as_str` away (it may be used at runtime if tracing is on) and it probably doesn't inline it (it's part of varargs). The end result is that we pay the price for all of the arguments to trace calls even if tracing is disabled.
> 
> In case of `as_str` this is exaggerated by the fact that the function will allocate internally (it uses std string formatting to construct the version string from numbers, one by one). We call this function a LOT (cca 700 times to run a simple console app). This internally will end up calling some of the string functions 4 times per call which is almost 3000 calls. All done for effectively nothing.

vitek-karas commented on src/corehost/common/trace.h:
```diff
@@ -31,6 +33,39 @@ namespace trace
+// verbosity level.
+#define TRACE_LAZY(Verbosity, Args...)                              \
+    do {                                                            \
+        if (Verbosity == trace::verbosity::Error)                   \
```
I personally would prefer to handle this difference in the macros below - basically have `TRACE_ERROR` implement itself fully and then have the others use the `TRACE_LAZY`.

lpereira replied:
Done.

vitek-karas commented on src/corehost/common/trace.h:
```diff
@@ -31,6 +33,39 @@ namespace trace
+        {                                                           \
+            /* Error conditions are rare: do not pollute the I$ */  \
+            /* real state with tracing calls. */                    \
+            auto cold_trace = [&]() COLD_FUNCTION NO_INLINE         \
```
I kind of like this - I assume it helps the compiler by moving the error related code out into a different method...
You may want to explain the `I$` term - I personally have no idea what it means (I can only guess from context).

AaronRobinsonMSFT replied:
Arbitrary lambda capture is incredibly expensive and the semantics aren't always obvious. If perf is a concern lambdas should be avoided entirely.

lpereira replied:
(`I$` is "instruction cache"; `$` (cash) sounds similar to "cache", and it's often used interchangeably in this context.  You might see things like `D$` for "data cache", or `L1D$` for the L1 data cache.)

AaronRobinsonMSFT commented on src/corehost/common/trace.h:
```diff
@@ -31,6 +33,39 @@ namespace trace
+// Tracing is performed with a macro so that arguments to trace::trace() and
+// trace::trace_error() can be evaluated lazily depending on the current
+// verbosity level.
+#define TRACE_LAZY(Verbosity, Args...)                              \
```
I would strongly avoid this pattern. It adds a lot to code bloat and isn't debuggable. Argument evaluation isn't a measurable cost here if all are pointer sized.

AaronRobinsonMSFT replied:
In this entire change there are no costly argument evaluations. Logging should never need to evaluate complex arguments so I am really confused as to this concern.

vitek-karas replied:
The suggestion came from profiling. During normal console app startup we do cca 700 `version_t.as_str()` which are pretty much all due to tracing. The method turns 4 numbers into a string a concatenates them - which is not exactly cheap.
I think that tracing might do semi-expensive things, that is turn structures to strings - so that it's humanly readable - this typically involves allocations and so on.
If you have an idea how to avoid it... I'm all for it. I looked on the internet and the recommendations are generally either macros (similar to this PR) or lambdas. There might be a solution based on C++ templates, but that seems to be rather tricky.

One question - why do you think this introduces a code bloat? All the macro does is add a simple if, which the compiler will very likely optimize to a simple memory read, comparison and branch.

AaronRobinsonMSFT replied:
Oh, I must have missed the `.as_str()`. I only saw a series of `.c_str()`. If that is the only scenario I would look for other solutions - templates specifically or in a pinch operator overloads for the classes in question.

Lambdas fine, but can have serious performance impact when used with capture - non-capture lambdas are generally fine.

The code bloat is related to black box that is a macro and inevitable additional code that will be inserted. There are patterns that are return value checks (i.e. `RETURN_IF_FAILED()`) which tend to be clear, but logging eventually turns into a grab bag of other features for additional logging and state recording during a crash or perhaps evaluation of asserts to aid in some other scenario. Basically logging and macros breed code bloat because "it's logging and we need more details".

This is based purely on my experience, but I think CoreCLR does a good job at demonstrating my concerns independent of my experience.

lpereira replied:
This actually reduces some of the code bloat.

The outermost `if` is evaluated at compile time (both sides in the comparisons are compile-time constants).  It could be a `if constexpr`, but we're not using C++17 yet.

In the non-error cases, calls to `current_verbosity()` should be hoisted -- it's marked as a pure function, so it should be just a single function call if the compiler decides not to inline it -- and that should be pretty much it: the rest will be either just integer comparisons, or, most likely, just one integer comparison for all the calls in the same verbosity within a function (due to constant propagation/folding).

Also, instead of having multiple versions of tracing functions, there's essentially just two: one for the error path, and other for all the others.  Having less code around is a good thing.

The error path is different, and it should be rare; that's why I put it into a lambda that's declared and called immediately (and marked as "noinline" where supported so it's far away from the happy path as far as things like branch predictors, prefetchers, uop caches and this sort of thing), capturing only by reference.  Most of the cases I saw were capturing a single `std::string.c_str()` or an integer type, so it should be fine.  I don't care much about the performance of the error path, as long as it doesn't interfere with the happy path -- that's why this idiom was used.

I've investigated other things, including using variadic templates.  I contemplated writing my own printf-like function that takes C++ types directly, and this remains an option here, considering that all tracing calls use very basic conversion specifiers (`%s`, `%d`, and `%x` -- with the occasional `%04x`). This could support `std::string` directly (removing the need to call `c_str()`, which makes the call sites cleaner), or arbitrary types with `operator std::string()` that could be implemented to perform expensive to-string operations only when necessary.  I decided to give the macro version a try first because it didn't require any changes to any of the calling sites that couldn't be done with an invocation of `sed` in the command line; but if people prefer this other method, I can try to come up with something.

AaronRobinsonMSFT replied:
>This actually reduces some of the code bloat.

My comment wasn't to be taken about present state, but precedent. Agreed that there are options to reduce using `constexpr`, but those aren't actually real because we don't want compile time optimization to exclude code because we want to enable logging on release builds as well so `constexpr` isn't something we can leverage - at least to my understanding.

>should be hoisted -- it's marked as a pure function, so it should be just a single function call if the compiler decides not to inline it 

Fully appreciate what the compiler can do, but again that isn't my concern. It is about precedence and what is needed. If logging is actually take more than 1 % when it is off then we are doing something wrong in general and playing optimization games tends to create fragile code paths that can easily become degenerate.

This leads into another principle - can most people look at this and understand how to maintain it? Comments on instructions caches tend to indicate code that relies on too many assumptions or deep knowledge about something that defeats the purpose of higher level languages. In this case we are talking about logging so if this is where we feel we need to hand optimize to that level then there should be literally no other hot spots in the code.

In the GC, JIT, type load, all of these comments and optimizations seem spot on. Not for logging which can be addressed with some basic templates that rely entirely on opaque compiler optimizations and not require any comments or descriptions about what the compiler can/should/may do.

> I decided to give the macro version a try first because it didn't require any changes to any of the calling sites that couldn't be done with an invocation of `sed` in the command line; but if people prefer this other method, I can try to come up with something.

Based on what I am seeing, I think the function should be convertible to a template without much code change, but of course that depends on how it is done. I would prefer to try something else if possible.

AaronRobinsonMSFT commented on src/corehost/common/trace.h:
```diff
@@ -31,6 +33,39 @@ namespace trace
+// trace::trace_error() can be evaluated lazily depending on the current
+// verbosity level.
+#define TRACE_LAZY(Verbosity, Args...)                              \
+    do {                                                            \
```
The `do` loop trick is only needed for C. For C++ just use scoping `{ ... }`

lpereira commented:
Thanks for the reviews.  I pushed a new revision of this PR, rebased on current HEAD.

It still uses the macros, but they were simplified by removing the outermost comparison. The self-calling lambda stayed, and I don't believe the implicit captures will cause any performance issue, given everything is captured by reference (and not copied around).  As a bonus, I also added another function attribute to the tracing functions taking a printf-like formatting string, which found a few bugs (that were fixed).

The reason I'm using macros and not using templates is that I tried a few different strategies yesterday and couldn't come up with anything that's cleaner than this.  I don't know yet C++ that well (I usually live in the C side of the pond), so I'm sure I missed some opportunities.  Happy to show what I've tried and why I didn't consider them to be better than these macros.  I don't usually like macros either, but there are times where they're the best tool for the job.

I've also changed a few things to make MSVC happier.

lpereira commented on src/corehost/common/trace.h:
```diff
@@ -9,16 +9,18 @@
 namespace trace
 {
+    enum class verbosity {
+        Disabled,
```
Done.

lpereira commented on src/corehost/common/trace.cpp:
```diff
@@ -45,81 +56,70 @@ bool trace::enable()
     pal::string_t tracefile_str;
 
-    if (g_trace_verbosity)
+    if (g_trace_verbosity != trace::verbosity::Disabled)
```
Done.

lpereira commented on src/corehost/common/trace.cpp:
```diff
@@ -6,17 +6,28 @@
 
 // g_trace_verbosity is used to encode COREHOST_TRACE and COREHOST_TRACE_VERBOSITY to selectively control output of
-//    trace::warn(), trace::info(), and trace::verbose()
+//    trace::warn(), TRACE_INFO(), and TRACE_VERBOSE()
```
Done.

vitek-karas commented on src/corehost/common/trace.cpp:
```diff
@@ -6,17 +6,23 @@
 
 // g_trace_verbosity is used to encode COREHOST_TRACE and COREHOST_TRACE_VERBOSITY to selectively control output of
-//    trace::warn(), trace::info(), and trace::verbose()
+//    trace::warn(), TRACE_INFO(), and TRACE_VERBOSE()
```
I still see `trace::warn` instead of `TRACE_WARNING` in the comment... can you please double check?

lpereira replied:
Oops. Done.

elinor-fung commented:
I know this shouldn't be a functional / logical change, but as @sdmaclea has pointed out in the past, there is a lot of caution and scrutiny over any changes affecting the muxer / `dotnet.exe`. We had promises to partners around not changing the muxer after preview 5, but we did actually change it in preview 6 to address some significant file size bloat. Is this change acceptable for us to do so late in the release?

Since the goal of the change is a performance win, I think we should definitely have some before and after measurements.

vitek-karas commented:
I absolutely agree with @elinor-fung  that we should measure the impact of this.
Regarding the `dotnet.exe` impact, I think we already have mechanisms how to deal with this:
* Thorough code-review - I explicitly asked about our story for the muxer and 3.0- RTM - my understanding is that we do expect to take a final build from 3.0 RTM and deploy that to our partners - they just asked that there are no functional changes since P5 and that they won't need to take any builds between P5 and RTM. But I might be wrong.
* If we really need to keep the muxer exactly the same, the build already has a way to specify an exact version of the "frozen" muxer - in which case building it will actually copy that version into the output without actually building it from the sources. AFAIK we will turn this on in servicing branches (and probably also 3.1) as a way to avoid updating the muxer and thus requiring a reboot when installing.

@jeffschwMSFT - to clarify our ability to make changes like this (which should have no functional impact) .

jeffschwMSFT commented:
At this point in the release we should not change the muxer.  @lpereira is this addressing a particular issue for .NET Core 3?  Based on the discussion, I think we should hold and put into our branch once we fork for the next release.

vitek-karas commented:
I think we should base the decision whether to consider it for 3.0, 3.1 or 5.0 based on data. Let's collect some perf numbers and see how much this helps... But I agree which should hold off merging this until then.

lpereira commented:
Absolutely -- no problems at all in waiting to merge this after 3.0 is released.

lpereira commented:
I have to think how I'm going to measure the impact of these changes.  It's not something that's easy to measure.  One of the things I can do, however, is look at the generated code and compare the two.

vitek-karas commented:
I think the measure problem goes back to what we've discussed previously - we want to compare wall clock time - that's really the only thing which will matter for deciding whether to take this change before 5 or not.  So measuring how long does it take for a process from start to "Some point after host" - this can be easier the time it gets into the C# Main, or have some other way to measure when it gets into CoreCLR.

sdmaclea commented:
If the time saved is less than 50ms it is unlikely to be noticed by the enduser. It is still worth the savings, but not for 3.0.

elinor-fung commented:
The more I think about introducing a bunch of macros (or templates) throughout the tracing system, the more apprehensive I am about making the system that much more complex and more annoying to debug to avoid a handful of `version_t::as_str` calls. Could we just do something like explicitly check `trace::is_enabled` before logging anything that might be expensive? 

I think we have maybe 10 or so cases where `version_t::as_str` is just used for logging. I tried profiling starting a console app on Windows - from what I saw, it really came down to about three of those taking up ~35% of the time. 

If we were to just explicitly check if trace is enabled before making any calls to `trace::info/verbose` that could be expensive, it would be targeted and easy to understand the impact of the change and still be a significant perf win.

We wouldn't have the 'magic' of not needing to worry about whether or not something just for logging might be expensive, but I actually kind of like the explicit expectation that we should be aware of the potential impact of anything we may compute for logging.

vitek-karas commented:
I guess I like that @elinor-fung . I'm a bit worried that eventually something will slip through, but then again that can be the case for any other code, not just logging. We need good perf tests to catch those kinds of things anyway.

I would be perfectly fine to surround some of the interesting bits of logging with an explicit ifs.

Funnily enough, that's exactly how it needs to be done in the CoreCLR, the tracing infra there also doesn't try to avoid argument evaluation automagically, instead it offers quick "is on" checks and expects callers to use those explicitly.

lpereira commented:
Alright; I'll simplify the patch so that it doesn't introduce any new macros.

dagood commented:
Is this PR still active?

lpereira commented:
Closing this as it's unlikely this'll get merged.  (There are probably better ways of improving this.)
